controllers/appgroup_controller.go

```diff
@@ -505,26 +525,63 @@ func (r *ApplicationGroupReconciler) cleanupWorkflow(ctx context.Context, logr l
 		for _, node := range wf.Status.Nodes {
 			nodes[node.ID] = node
 		}
-		graph, err := workflow.Build(g.Name, nodes)
+		rwf := &v1alpha12.Workflow{}
+
+		rwfName := fmt.Sprintf("%s-reverse", wf.Name)
+		rwfNamespace := wf.Namespace
+		err := r.Client.Get(ctx, types.NamespacedName{Namespace: rwfNamespace, Name: rwfName}, rwf)
 		if err != nil {
-			logr.Error(err, "failed to build the wf status DAG")
-		}
+			if kerrors.IsNotFound(err) {
+				err = r.generateReverseWorkflow(ctx, logr, nodes, &g)
+				if err != nil {
+					logr.Error(err, "failed to generate reverse workflow")
+					return err
+				}
 
-		rev := graph.Reverse()
+				// update the forward workflow metadata with a finalizer
+				wf.Finalizers = []string{rwffinalizer}
 
-		for _, bucket := range rev {
-			for _, hr := range bucket {
-				err = r.Client.Delete(ctx, &hr)
+				wfPatch := client.MergeFrom(wf.DeepCopy())
+				_ = r.Client.Patch(ctx, &wf, wfPatch)
+
+				err = r.Client.Delete(ctx, &wf)
 				if err != nil {
-					logr.Error(err, "failed to delete helmrelease CRO - continuing with cleanup")
+					logr.Error(err, "failed to delete workflow CRO - continuing with cleanup")
+					return err
 				}
+
+				// requeue
+				return ErrStartedReverseWorkflow
 			}
-		}
+		} else {
+			// check the completion of the reverse workflow
+			if !rwf.Status.FinishedAt.IsZero() {
+				// remove the finalizer from the forward workflow
+				wf.Finalizers = nil
```

Comment:
In general, you shouldn't set finalizers to `nil` because it's possible there are other finalizers on the object that other things are waiting on. Can you use `controllerutil.RemoveFinalizer()` instead?

Reply:
Actually I was looking for this and even almost wrote my own RemoveFinalizer. Thanks.